services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/NeutronRouterServiceImpl.java

```diff
@@ -469,6 +468,197 @@ public ConnectedSubnetsWebResponse getConnectedSubnets(String projectId, String
         return connectedSubnetsWebResponse;
     }
 
+    @Override
+    public UpdateRoutingRuleResponse updateRoutingRule (String owner, NewRoutesWebRequest newRouteEntry, boolean isNeutronOrVPCLevelRoutingRule) throws DestinationOrNexthopCanNotBeNull, CacheException, CanNotFindRouteTableByOwner, QueryParamTypeNotSupportException, RouteTableNotUnique, DestinationInvalid {
+        List<InternalRoutingRule> updateRoutes = new ArrayList<>();
+        List<HostRoute> hostRouteToSubnet = new ArrayList<>();
+
+        List<NewRoutesRequest> routes = newRouteEntry.getRoutes();
+        if (routes == null) {
+            return new UpdateRoutingRuleResponse(new ArrayList<>(), hostRouteToSubnet);
+        }
+
+        // find routeTable
+        Map<String, String[]> requestParams =  new HashMap<>();
+        String[] value = new String[1];
+        value[0] = owner;
+        requestParams.put("owner", value);
+
+        Map<String, Object[]> queryParams =
+                ControllerUtil.transformUrlPathParams(requestParams, RouteTable.class);
+        Map<String, RouteTable> routeTableMap = this.routeTableDatabaseService.getAllRouteTables(queryParams);
+        List<RouteTable> routeTables = new ArrayList<>(routeTableMap.values());
+        if (routeTables == null || routeTables.size() == 0) {
+            logger.log(Level.WARNING, "owner: " + owner);
+            throw new CanNotFindRouteTableByOwner();
+        } else if (routeTables.size() >= 2) {
+            logger.log(Level.WARNING, "owner: " + owner);
+            throw new RouteTableNotUnique();
+        }
+
+        RouteTable existRouteTable = routeTables.get(0);
+
+        List<RouteEntry> existRoutes = existRouteTable.getRouteEntities();
+        // TODO: existRoutes -> MAP: key - des(/前面的), value - nexthop, "10.0.0.0/16"
+        Map<String, RouteEntry> existRoutesMap = new HashMap<>();
+        for (RouteEntry existRoute : existRoutes) {
+            String[] existDes = existRoute.getDestination().split("\\/");
+            String existNetworkIP = existDes[0];
+            existRoutesMap.put(existNetworkIP, existRoute);
+        }
+        // TODO: existRoutes replaced with map
+        // Tracking operation type for each routing rule
+        for (NewRoutesRequest newRouteRequest : routes) {
+            //RouteEntry route = null;
+            // TODO: 1. check new-des if it is valid
+            String newRouteDestination = newRouteRequest.getDestination();
+            String newRouteNexthop = newRouteRequest.getNexthop();
+            boolean isDestinationValid = verifyCidrBlock(newRouteDestination);
+            if (!isDestinationValid) {
+                throw new DestinationInvalid("destination is invalid : " + newRouteDestination);
+            }
+            String[] newRouteDes = newRouteDestination.split("\\/");
+            String newNetworkIP = newRouteDes[0];
+            String newBitmask = newRouteDes[1];
+            int newBitmaskInt = Integer.parseInt(newBitmask);
```

Comment:
Do we need put these operations here? or move them into `else` part below if the `if` part doesn't need them.